Common/DataMonitor.cs

```diff
@@ -0,0 +1,265 @@
+                }
+            })
+            { IsBackground = true };
+            _requestRateCalculationThread.Start();
+        }
+        
+        private DataMonitorReport GenerateReport()
+        {
+            var report = new DataMonitorReport(_succeededDataRequestsCount, 
+                _failedDataRequestsCount, 
+                _succeededUniverseDataRequestsCount, 
+                _failedUniverseDataRequestsCount, 
+                _requestRates);
+
+            Logging.Log.Trace($"DataMonitor.GenerateReport():{Environment.NewLine}" +
+                $"DATA USAGE:: Total data requests {report.TotalRequestsCount}{Environment.NewLine}" +
+                $"DATA USAGE:: Succeeded data requests {report.SucceededDataRequestsCount}{Environment.NewLine}" +
+                $"DATA USAGE:: Failed data requests {report.FailedDataRequestsCount}{Environment.NewLine}" +
+                $"DATA USAGE:: Failed data requests percentage {report.FailedDataRequestsPercentage}%{Environment.NewLine}" +
+                $"DATA USAGE:: Total universe data requests {report.TotalUniverseDataRequestsCount}{Environment.NewLine}" +
+                $"DATA USAGE:: Succeeded universe data requests {report.SucceededUniverseDataRequestsCount}{Environment.NewLine}" +
+                $"DATA USAGE:: Failed universe data requests {report.FailedUniverseDataRequestsCount}{Environment.NewLine}" +
+                $"DATA USAGE:: Failed universe data requests percentage {report.FailedUniverseDataRequestsPercentage}%");
+
+            return report;
+        }
+
+        private void ComputeFileRequestFrequency()
+        {
+            var requestsCount = _succeededDataRequestsCount + _failedDataRequestsCount;
+
+            if (_lastRequestRateCalculationTime == default)
+            {
+                // First time we calculate the request rate.
+                // We don't have a previous value to compare to so we just store the current value.
+                _lastRequestRateCalculationTime = DateTime.UtcNow;
+                _prevRequestsCount = requestsCount;
+                return;
+            }
+
+            var requestsCountDelta = requestsCount - _prevRequestsCount;
+            var now = DateTime.UtcNow;
+            var timeDelta = now - _lastRequestRateCalculationTime;
+
+            _requestRates.Add(Math.Round(requestsCountDelta / timeDelta.TotalSeconds));
+            _prevRequestsCount = requestsCount;
+            _lastRequestRateCalculationTime = now;
+        }
+
+        /// <summary>
+        /// Stores the data monitor report
+        /// </summary>
+        /// <param name="report">The data monitor report to be stored<param>
+        private void StoreDataMonitorReport(DataMonitorReport report)
+        {
+            if (report == null)
+            {
+                return;
+            }
+
+            var path = GetFilePath("data-monitor-report.json");
+            var data = JsonConvert.SerializeObject(report, Formatting.None  );
```

Comment:
`Formatting.None  );` nit spaces

`GetFilePath` is adding the timestamp after the file extension, ideally it's before it

Reply:
My  bad with the timestamp, it has to be before the tghe extension 🎯